### PR TITLE
[WIP][IRGenDebugInfo] Rewrite the line entry verifier.

### PR DESCRIPTION
**Pull request by dcci**
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

### CODE REVIEW

**Comment by dcci** [comment]
@swift-ci test

**Comment by dcci** [comment]
Not really ready for review yet. I'm trying to figure out whether the algorithm works looking at the failures it triggered. The previous one had flaws I'm going to describe soon.

**Comment by dcci** [comment]
OK, yes, looks like this one finds real bugs.

```
sil_scope 1 { loc "a.swift":3:13 parent @$s1a1fyySiSgF : $@convention(thin) (Optional<Int>) -> () }
sil_scope 2 { loc "a.swift":4:1 parent 1 }
// %4                                             // users: %9, %5
bb1(%4 : $Int):                                   // Preds: bb0
  store %4 to %2 : $*Int, loc "a.swift":6:18, scope 2 // id: %5
  %6 = begin_access [read] [static] %2 : $*Int, loc "a.swift":7:12, scope 2 // user: %7
  end_access %6 : $*Int, loc "a.swift":7:12, scope 2 // id: %7
  %8 = alloc_stack $Int, loc "a.swift":7:12, scope 2 // users: %9, %12, %11
  store %4 to %8 : $*Int, loc "a.swift":7:12, scope 2 // id: %9
  // function_ref markUsed<A>(_:)
  %10 = function_ref @$s1a8markUsedyyxlF : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0) -> (), loc "a.swift":7:3, scope 2 // user: %11
  %11 = apply %10<Int>(%8) : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0) -> (), loc "a.swift":7:3, scope 2
  dealloc_stack %8 : $*Int, loc "a.swift":7:14, scope 2 // id: %12
  dealloc_stack %2 : $*Int, loc "a.swift":6:13, scope 2 // id: %13
  br bb3, loc "a.swift":8:1, scope 2              // id: %14
```

The latest `dealloc_stack` points to line 6 and the previous one points to line 7 in straight line code. It's a jump backwards, which shouldn't happen.

[taken from `test/DebugInfo/GuardVar.swift` compiled with `-g -Xllvm -verify-linetable`

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 1]
```diff
@@ -127,17 +127,18 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
       LocationStack;
 
 #ifndef NDEBUG
-  using UUSTuple = std::pair<std::pair<unsigned, unsigned>, StringRef>;
+  using UUSTuple = std::pair<unsigned, StringRef>;
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp]: I suppose the old name meant unsigned-unsigned-string tuple; now the name doesn't make sense any more. Is there a better name for it?

Comment (adrian-prantl) [reply lib/IRGen/IRGenDebugInfo.cpp]: maybe just don't use the using decl at all

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 2]
```diff
@@ -1759,16 +1760,25 @@ bool IRGenDebugInfoImpl::lineEntryIsSane(SILLocation::DebugLoc DL,
...
-  if (DL.Line == 0 || DL == PreviousDebugLoc)
+    PreviousDebugLoc = DebugLocKey();
+
+  // Skip line entry zero.
+  if (DL.Line == 0)
+    return true;
+
+  auto DLK = DebugLocKey(DL);
+
+  if (DLK >= PreviousDebugLoc || !DLK.sameFile(PreviousDebugLoc.second)) {
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp]: Can you add a comment that explains that this verifies that line numbers are monotonically increasing within basic blocks?

**Comment by dcci** [comment]
Superseded by https://github.com/apple/swift/pull/26764

**Pull request closed without merging**